### PR TITLE
Master settings fiximp tde

### PR DESCRIPTION
**Pull request by tde-banana-odoo**
Task: https://www.odoo.com/web#id=32196&view_type=form&model=project.task&action=333&active_id=965&menu_id=4720

### CODE REVIEW

File: addons/website/models/res_config.py [hunk 1]
```diff
@@ -71,14 +70,33 @@ def _compute_language_count(self):
         for config in self:
             config.language_count = len(self.language_ids)
 
-    def set_has_google_analytics(self):
-        return self.env['ir.values'].set_default(
-            'website.config.settings', 'has_google_analytics', self.has_google_analytics)
+    @api.model
```
Comment (xmo-odoo) [inline addons/website/models/res_config.py]: Are these decorators actually useful/necessary?

File: addons/mrp/models/mrp_config_settings.py [hunk 2]
```diff
@@ -22,3 +22,18 @@ class MrpConfigSettings(models.TransientModel):
...
+        )
+
+    @api.multi
+    def set_fields(self):
+        self.env['ir.config_parameter'].sudo().set_param('mrp.use_manufacturing_lead', self.use_manufacturing_lead)
+
+    @api.onchange('use_manufacturing_lead')
+    def _onchange_use_manufacturing_lead(self):
+        if not self.use_manufacturing_lead:
+            self.manufacturing_lead = 0.0
```
Comment (Icallhimtest) [inline addons/mrp/models/mrp_config_settings.py]: Missing end of line

File: addons/point_of_sale/models/res_config.py [hunk 3]
```diff
@@ -56,3 +57,13 @@ def _onchange_sale_price(self):
                 'group_sale_pricelist': False,
                 'group_pricelist_item': False,
             })
+
+    def get_default_fields(self, fields):
+        return dict(
```
Comment (Icallhimtest) [inline addons/point_of_sale/models/res_config.py]: Use dict literal if you please.

File: addons/hr_expense/models/hr_expense_config_settings.py [hunk 4]
```diff
@@ -21,10 +21,12 @@ class HrExpenseConfigSettings(models.TransientModel):
 
     @api.model
     def get_default_fields(self, fields):
-        return dict(
+        res = super(HrExpenseConfigSettings, self).get_default_fields(fields)
+        res.update(dict(
```
Comment (rco-odoo) [inline addons/hr_expense/models/hr_expense_config_settings.py]: `dict` is not necessary, you can use `update()` with kwargs.

File: addons/mrp/models/mrp_config_settings.py [hunk 5]
```diff
@@ -25,9 +25,11 @@ class MrpConfigSettings(models.TransientModel):
 
     @api.model
     def get_default_fields(self, fields):
-        return dict(
+        res = super(MrpConfigSettings, self).get_default_fields(fields)
+        res.update(dict(
```
Comment (rco-odoo) [inline addons/mrp/models/mrp_config_settings.py]: `dict` is not necessary, you can use `update()` with kwargs.

File: addons/point_of_sale/models/pos_config_settings.py [hunk 6]
```diff
@@ -59,10 +59,12 @@ def _onchange_sale_price(self):
             })
 
     def get_default_fields(self, fields):
-        return dict(
+        res = super(PosConfiguration, self).get_default_fields()
+        res.update(dict(
```
Comment (rco-odoo) [inline addons/point_of_sale/models/pos_config_settings.py]: `dict` is not necessary, you can use `update()` with kwargs.

File: addons/purchase/models/stock_config_settings.py [hunk 7]
```diff
@@ -19,9 +19,11 @@ def _onchange_use_po_lead(self):
             self.po_lead = 0.0
 
     def get_default_fields(self, fields):
-        return dict(
+        res = super(StockConfigSettings, self).get_default_fields(fields)
+        res.update(dict(
```
Comment (rco-odoo) [inline addons/purchase/models/stock_config_settings.py]: `dict` is not necessary, you can use `update()` with kwargs.

File: addons/sale/models/sale_config_settings.py [hunk 8]
```diff
@@ -148,11 +148,13 @@ def set_multi_sales_price(self):
 
     @api.model
     def get_default_fields(self, fields):
-        return dict(
+        res = super(SaleConfiguration, self).get_default_fields(fields)
+        res.update(dict(
```
Comment (rco-odoo) [inline addons/sale/models/sale_config_settings.py]: `dict` is not necessary, you can use `update()` with kwargs.

File: addons/sale_stock/models/stock_config_settings.py [hunk 9]
```diff
@@ -23,9 +23,11 @@ def _onchange_use_security_lead(self):
             self.security_lead = 0.0
 
     def get_default_fields(self, fields):
-        return dict(
+        res = super(StockConfigSettings, self).get_default_fields(fields)
+        res.update(dict(
```
Comment (rco-odoo) [inline addons/sale_stock/models/stock_config_settings.py]: `dict` is not necessary, you can use `update()` with kwargs.

File: addons/stock/models/stock_config_settings.py [hunk 10]
```diff
@@ -76,9 +76,11 @@ def onchange_adv_location(self):
 
     @api.model
     def get_default_fields(self, fields):
-        return dict(
+        res = super(StockConfigSettings, self).get_default_fields(fields)
+        res.update(dict(
```
Comment (rco-odoo) [inline addons/stock/models/stock_config_settings.py]: `dict` is not necessary, you can use `update()` with kwargs.

File: addons/website/models/website_config_settings.py [hunk 11]
```diff
@@ -89,11 +89,13 @@ def _compute_language_count(self):
     def get_default_fields(self, fields):
         if not self.user_has_groups('website.group_website_designer'):
             raise AccessDenied()
-        return dict(
+        res = super(WebsiteConfigSettings, self).get_default_fields(fields)
+        res.update(dict(
```
Comment (rco-odoo) [inline addons/website/models/website_config_settings.py]: `dict` is not necessary, you can use `update()` with kwargs.

File: odoo/addons/base/res/res_config.py [hunk 12]
```diff
@@ -504,9 +507,11 @@ def default_get(self, fields):
                 res[name] = int(res[name])
 
         # other fields: call all methods that start with 'get_default_'
+        # and call the method 'get_default_fields'
         for method in dir(self):
```
Comment (rco-odoo) [inline odoo/addons/base/res/res_config.py]: This `for` loop may be deprecated in favor of the new single method `get_default_fields`...
(You may log a warning when a getter method is actually called this way.)

File: odoo/addons/base/res/res_config.py [hunk 13]
```diff
@@ -545,9 +548,11 @@ def execute(self):
                 implied_group.write({'users': [(3, user.id) for user in groups.mapped('users')]})
 
         # other fields: execute all methods that start with 'set_'
+        # and the method 'set_fields'
         for method in dir(self):
```
Comment (rco-odoo) [inline odoo/addons/base/res/res_config.py]: Like with method `get_default_fields`, this `for` loop may be deprecated in favor of the single method `set_fields`...
(You may log a warning if setter methods are actually called this way.)

File: odoo/addons/base/res/res_config.py (hunk 12 above)
Comment (tivisse) [inline odoo/addons/base/res/res_config.py]: Good idea indeed. :)

File: odoo/addons/base/res/res_config.py [hunk 14]
```diff
@@ -478,6 +458,9 @@ def _get_classified_fields(self):
 
         return {'default': defaults, 'group': groups, 'module': modules, 'other': others}
 
+    def get_default_fields(self, fields):
```
Comment (rco-odoo) [inline odoo/addons/base/res/res_config.py]: The method name and API is not so good, IMHO:
* Why "default" fields? What you ask for is the current value of the stuff, shown as fields in the model. The values are fed as default values in the wizard, but that's an implementation trick.
* Why "fields"? What you ask for are configuration parameters, and such things.
* Why passing a list of fields? Its value is *never* used.

I would further simplify the API of both methods to something like
```python
def get_values(self):
    return {}

def set_values(self):
    pass
```

File: odoo/addons/base/module/module.py [hunk 15]
```diff
@@ -479,6 +479,20 @@ def upstream_dependencies(self, known_deps=None,
             known_deps |= missing_mods.upstream_dependencies(known_deps, exclude_states)
         return known_deps
 
+    def next(self):
+        Todos = self.env['ir.actions.todo']
+        _logger.info('getting next %s', Todos)
+        active_todo = Todos.search([('state', '=', 'open')], limit=1)
+        if active_todo:
+            _logger.info('next action is %s', active_todo)
+        if active_todo:
```
Comment (rco-odoo) [inline odoo/addons/base/module/module.py]: both `if`'s may be merged together

File: odoo/addons/base/module/module.py [hunk 16]
```diff
@@ -479,6 +479,20 @@ def upstream_dependencies(self, known_deps=None,
             known_deps |= missing_mods.upstream_dependencies(known_deps, exclude_states)
         return known_deps
 
+    def next(self):
+        Todos = self.env['ir.actions.todo']
```
Comment (rco-odoo) [inline odoo/addons/base/module/module.py]: add a bit of doc, please

File: addons/account/models/account_config_settings.py [hunk 17]
```diff
@@ -60,34 +60,28 @@ class AccountConfigSettings(models.TransientModel):
...
 
     @api.model
-    def get_default_tax_fields(self, fields):
-        default_purchase_tax_id = self.env['ir.config_parameter'].sudo().get_param('account.default_purchase_tax_id', default=False)
-        default_sale_tax_id = self.env['ir.config_parameter'].sudo().get_param('account.default_sale_tax_id', default=False)
-        return dict(default_purchase_tax_id=int(default_purchase_tax_id), default_sale_tax_id=int(default_sale_tax_id))
+    def get_values(self):
+        res = super(AccountConfigSettings, self).get_values()
+        params = self.env['ir.config_parameter'].sudo()
+        res.update(dict(
```
Comment (rco-odoo) [inline addons/account/models/account_config_settings.py]: simplify: `res.update(dict(...))` -> `res.update(...)`

File: odoo/addons/base/res/res_config.py [hunk 18]
```diff
@@ -478,6 +458,9 @@ def _get_classified_fields(self):
 
         return {'default': defaults, 'group': groups, 'module': modules, 'other': others}
 
+    def get_values(self):
+        return {}
```
Comment (rco-odoo) [inline odoo/addons/base/res/res_config.py]: document: "Return values for the fields other than defaults etc. "

File: odoo/addons/base/res/res_config.py [hunk 19]
```diff
@@ -503,13 +486,18 @@ def default_get(self, fields):
...
         for method in dir(self):
             if method.startswith('get_default_'):
-                res.update(getattr(self, method)(fields))
+                _logger.warning(_('Methods that start with `get_default_` are deprecated. Override `get_values` instead(Method %s)') % method)
+        res.update(self.get_values())
 
         return res
 
+    def set_values(self):
+        pass
```
Comment (rco-odoo) [inline odoo/addons/base/res/res_config.py]: document: "Set values for the fields other than defaults etc. "

**Pull request merged**